lms/static/scripts/frontend_apps/components/test/BookPicker-test.js

```diff
@@ -0,0 +1,142 @@
+import { mount } from 'enzyme';
+import { createElement } from 'preact';
+import { act } from 'preact/test-utils';
+
+import mockImportedComponents from '../../../test-util/mock-imported-components';
+import { waitForElement } from '../../../test-util/wait';
+import * as bookData from '../../utils/vitalsource-sample-data';
+import BookPicker, { $imports } from '../BookPicker';
+
+describe('BookPicker', () => {
+  let fakeAPI;
+
+  const renderBookPicker = (props = {}) =>
+    mount(<BookPicker authToken="dummy-auth-token" {...props} />);
+
+  beforeEach(() => {
+    fakeAPI = {
+      fetchBooks: sinon.stub().resolves(bookData.bookList),
+      fetchChapters: sinon
+        .stub()
+        .callsFake(async (authToken, bookID) => bookData.chapterData[bookID]),
+    };
+
+    $imports.$mock(mockImportedComponents());
+    $imports.$restore({
+      './Dialog': true,
+    });
+    $imports.$mock({
+      '../utils/api': fakeAPI,
+    });
+  });
+
+  afterEach(() => {
+    $imports.$restore();
+  });
+
+  const selectBook = wrapper => {
+    const bookList = wrapper.find('BookList');
+    const book = bookList.prop('books')[0];
+    act(() => {
+      bookList.prop('onSelectBook')(book);
+    });
+    wrapper.update();
+    return book;
+  };
+
+  const selectChapter = wrapper => {
+    const chapterList = wrapper.find('ChapterList');
+    const chapter = chapterList.prop('chapters')[0];
+    act(() => {
+      chapterList.prop('onSelectChapter')(chapter);
+    });
+    wrapper.update();
+    return chapter;
+  };
+
+  const clickSelectButton = wrapper => {
+    act(() => {
+      wrapper
+        .find('LabeledButton[data-testid="select-button"]')
+        .prop('onClick')();
+    });
+    wrapper.update();
+  };
+
+  it('fetches and displays book list when picker is opened', async () => {
+    const picker = renderBookPicker();
+
+    let bookList = picker.find('BookList');
+    assert.deepEqual(bookList.prop('books'), []);
+    assert.isTrue(bookList.prop('isLoading'));
+    assert.calledWith(fakeAPI.fetchBooks, 'dummy-auth-token');
+
+    await waitForElement(picker, 'BookList[isLoading=false]');
+
+    bookList = picker.find('BookList');
+    assert.deepEqual(bookList.prop('books'), bookData.bookList);
+  });
+
+  it('fetches and displays chapter list when a book is chosen', async () => {
+    const picker = renderBookPicker();
+    await waitForElement(picker, 'BookList[isLoading=false]');
+
+    const book = selectBook(picker);
+    clickSelectButton(picker);
+
+    assert.isFalse(picker.exists('BookList'));
+    let chapterList = picker.find('ChapterList');
+    assert.isTrue(chapterList.exists());
+    assert.equal(chapterList.prop('isLoading'), true);
+    assert.calledWith(fakeAPI.fetchChapters, 'dummy-auth-token');
+
+    await waitForElement(picker, 'ChapterList[isLoading=false]');
+
+    chapterList = picker.find('ChapterList');
+    assert.equal(chapterList.prop('chapters'), bookData.chapterData[book.id]);
+  });
```

Comment:
This set of tests also tests the selection callback (i.e. what happens when a chapter is selected) in addition to what its test description captures. i.e perhaps there is more than one test here?

Reply:
This test does check what happens when a book is selected, but that's necessary to get the chapter list to appear in a way that realistically simulates the user. It doesn't check what happens when a chapter is selected though. That's in the next test.